rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -3193,6 +3375,18 @@ extern "C" rmw_ret_t rmw_trigger_guard_condition(
   RET_WRONG_IMPLID(guard_condition_handle);
   auto * gcond_impl = static_cast<CddsGuardCondition *>(guard_condition_handle->data);
   dds_set_guardcondition(gcond_impl->gcondh, true);
+
+  // Hack: Get and call the guard condition's listener callback,
+  // as it doesn't seem to be listening to the proper enitity.
```

Comment:
From a quick look I think that the problem is that the `dds_entity_t` listener field (named `m_listener`) is only used in the `dds_reader_data_available_cb` function.

I think that it is not an hack to do that here, but there are some changes required.
 - you should check the return type of `dds_set_guardcondition`
 - you should do the whole procedure that is currently done in `dds_reader_data_available_cb`
 
 something like this
 ```
 struct dds_listener const * const lst = &rd->m_entity.m_listener;
  dds_entity * const sub = rd->m_entity.m_parent;
  if (lst->on_data_on_readers)
  {
    ddsrt_mutex_unlock (&rd->m_entity.m_observers_lock);
    ddsrt_mutex_lock (&sub->m_observers_lock);
    const uint32_t data_on_rds_enabled = (ddsrt_atomic_ld32 (&sub->m_status.m_status_and_mask) & (DDS_DATA_ON_READERS_STATUS << SAM_ENABLED_SHIFT));
    if (data_on_rds_enabled)
    {
      sub->m_cb_pending_count++;
      while (sub->m_cb_count > 0)
        ddsrt_cond_wait (&sub->m_observers_cond, &sub->m_observers_lock);
      sub->m_cb_count++;
      ddsrt_mutex_unlock (&sub->m_observers_lock);

      lst->on_data_on_readers (sub->m_hdllink.hdl, lst->on_data_on_readers_arg);

      ddsrt_mutex_lock (&sub->m_observers_lock);
      sub->m_cb_count--;
      sub->m_cb_pending_count--;
      ddsrt_cond_broadcast (&sub->m_observers_cond);
    }
    ddsrt_mutex_unlock (&sub->m_observers_lock);
    ddsrt_mutex_lock (&rd->m_entity.m_observers_lock);
  }
  else if (lst->on_data_available)
  {
    ddsrt_mutex_unlock (&rd->m_entity.m_observers_lock);
    lst->on_data_available (rd->m_entity.m_hdllink.hdl, lst->on_data_available_arg);
    ddsrt_mutex_lock (&rd->m_entity.m_observers_lock);
  }
  ```

Reply:
I have to dig a little more on this, but I think that if you're allowed to attach a listener to an entity, whatever happens with that entity that changes the state of it, should call the listener callback

Reply:
And yes, seems that `m_listener` is only used in the `dds_reader_data_available_cb` function.
I'll try to implement it here and then we can ask the guys if that's the correct approach

Reply:
Well seems not straight forward to implement it here. There are no readers, the guard condition is just a single entity.
What I can do is lock the mutex of the guard condition as done in `dds_set_guardcondition` , get the callback, copy it to a new one. Unlock, and call the new one. 
Or just leave it as it is now and ask for guidance?